### PR TITLE
Bluetooth Patches to enable HFP

### PR DESCRIPTION
**Pull request by aiswaryacyriac**
Tracked-On: https://jira.devtools.intel.com/browse/OAM-75510
Signed-off-by: Aiswarya Cyriac <aiswarya.cyriac@intel.com>

### CODE REVIEW

**Comment by aiswaryacyriac** [comment]
@JeevakaPrabu

**Comment by aiswaryacyriac** [comment]
@Laxminarayan Bharadiya, Pankaj

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 1]
```diff
@@ -0,0 +1,1112 @@
...
+new file mode 100644
+index 000000000..2a5706785
+--- /dev/null
++++ b/drivers/misc/btusb_sco_snd_card.c
+@@ -0,0 +1,1030 @@
++/*
++ *
++ *  Bluetooth SCO driver
++ *
++ *  Copyright (C) 2019-2020 Intel Corporation
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: Please use license header as mentioned in below link and add all sign-off as author names.

https://elixir.bootlin.com/linux/v5.1-rc3/source/sound/soc/soc-compress.c

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 2]
```diff
@@ -0,0 +1,1112 @@
...
++
++#include <sound/core.h>
++#include <sound/info.h>
++#include <sound/pcm.h>
++
++#include <linux/usb.h>
++#include <linux/usbdevice_fs.h>
++#include <asm/byteorder.h>
++
++
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: Remove extra lines..

Try to arrange header alphabetically.

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 3]
```diff
@@ -0,0 +1,1112 @@
...
++#define MAX_PACKET_SIZE		(SCO_HDR_SIZE + MIN_PERIOD_SIZE)
++#define BTUSB_MAX_ISOC_FRAMES	10
++
++#define BTUSB_SCO_ALTERNATE_SETTING	2
++
++#define BTUSB_ISOC_TX_EP_CONFIGURED 1
++#define BTUSB_ISOC_RX_EP_CONFIGURED 2
++#define BTUSB_ISOC_TX_START	    3
++#define BTUSB_ISOC_RX_START	    4
++#define MAX_URBS		    12
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: Make all tab spaces uniform.

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 4]
```diff
@@ -0,0 +1,1112 @@
...
++	unsigned int playback_transfer_done;	///* processed frames since last period update */
++	unsigned int capture_hwptr_done;
++	unsigned int capture_transfer_done;
++
++	struct usb_anchor rx_anchor;
++
++	struct urb *tx_urb;
++	struct urb *rx_urb[MAX_URBS];
++};
++
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: Move struct definition before function definitions.

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 5]
```diff
@@ -0,0 +1,1112 @@
...
++MODULE_DEVICE_TABLE(usb, btusb_sco_table);
++
++static void dummy_proc_init(struct snd_card *card)
++{
++	struct snd_info_entry *entry;
++	if (!snd_card_proc_new(card, "dummy_pcm", &entry)) {
++		printk("btusb_sco:Registred sounc card in proc");
++	}
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: Do we really need this?  Please remove it.

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 6]
```diff
@@ -0,0 +1,1112 @@
...
++	{ USB_DEVICE(0x8087, 0x0a2b), .driver_info = BTUSB_INTEL_NEW },
++	{ USB_DEVICE(0x8087, 0x0aaa), .driver_info = BTUSB_INTEL_NEW },
++	{ USB_DEVICE(0x8087, 0x0aa7), .driver_info = BTUSB_INTEL_NEW },
++	{ }	/* Terminating entry */
++};
++
++MODULE_DEVICE_TABLE(usb, btusb_sco_table);
++
++static void dummy_proc_init(struct snd_card *card)
++{
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: Why this function is needed?

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 7]
```diff
@@ -0,0 +1,1112 @@
...
++	struct snd_info_entry *entry;
++	if (!snd_card_proc_new(card, "dummy_pcm", &entry)) {
++		printk("btusb_sco:Registred sounc card in proc");
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: remove printk, use dev_dbg.., dev_* APIs for prints

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 8]
```diff
@@ -0,0 +1,1112 @@
...
++
++	urb->number_of_packets = i;
++}
++
++static int btusb_isoc_initialize_tx_urb(struct btusb_data *data)
++{
++	unsigned int pipe;
++
++	if (!data->isoc_tx_ep) {
++		pr_emerg("%s - isoc_tx_ep is NULL\n", __func__);
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: remove pr_emerg calls from entire code use dev_dbg, dev_err , dev_* APIs instead

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 9]
```diff
@@ -0,0 +1,1112 @@
...
++	return 0;
++}
++
++static void btusb_isoc_prepare_tx_urb(struct btusb_data *data)
++{
++	unsigned int stride, frames, bytes;
++	static unsigned int oldptr;
++	uint8_t period_elapsed = 0;
++	unsigned long flags;
++	uint8_t *hw_ptr;
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: hw_ptr name looks misleading here. Please rename it to something better.. may be sco_buf or

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 10]
```diff
@@ -0,0 +1,1112 @@
...
++	oldptr = data->playback_hwptr_done;
++	if (data->playback_hwptr_done >=
++			data->playback_stream->runtime->buffer_size * stride)
++		data->playback_hwptr_done -=
++			data->playback_stream->runtime->buffer_size * stride;
++
++	memcpy(hw_ptr + SCO_HDR_SIZE,
++	       data->playback_stream->runtime->dma_area + data->playback_hwptr_done,
++	       bytes);
++
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: Logic looks incorrect. We need to accommodate of a case where we have few bytes available at the end of "dma_area" ring buffer which is not sufficient to accommodate entire sco buffer. In such case we need to copy part of data at the beginning of ring buffer.

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 11]
```diff
@@ -0,0 +1,1112 @@
...
++		return -ENODEV;
++	}
++
++	return 0;
++}
++
++#define BUFF_SIZE_MAX	(PAGE_SIZE * 16)
++#define PRD_SIZE_MAX	PAGE_SIZE  /*4096*/
++#define MIN_PERIODS	4
++
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: Move  macros at the beginning of file.

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 12]
```diff
@@ -0,0 +1,1112 @@
...
++static struct snd_pcm_hardware btsco_pcm_hardware = {
++	.info = SNDRV_PCM_INFO_INTERLEAVED | SNDRV_PCM_INFO_BLOCK_TRANSFER //field contains the type and capabilities of this pcm
++		| SNDRV_PCM_INFO_MMAP | SNDRV_PCM_INFO_MMAP_VALID,
++	.rates = SNDRV_PCM_RATE_CONTINUOUS,
++	.buffer_bytes_max = BUFF_SIZE_MAX,
++	.period_bytes_min = MIN_PERIOD_SIZE,
++	.period_bytes_max = 16384,
++	.periods_min =	    MIN_PERIODS,
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: Are we supporting 16384 period bytes
?

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 13]
```diff
@@ -0,0 +1,1112 @@
...
++static int bt_pcm_hw_params(struct snd_pcm_substream *substream,
++			     struct snd_pcm_hw_params *hw_params)
++{
++	struct btusb_data *data;
++	int err;
++	unsigned long flags;
++	data = snd_pcm_substream_chip(substream);
++	err = snd_pcm_lib_malloc_pages(substream,
++			params_buffer_bytes(hw_params));
++
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: Need to have check for supported PCM config. -> format, period size, etc checks,

Will driver work if user-space provides arbitrary period size?

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 14]
```diff
@@ -0,0 +1,1112 @@
...
++			sizeof(card->longname));
++
++	err = snd_card_register(card);
++	if (err){
++		goto register_fail;
++	} else {
++		printk("%s:Sound card registred sucecssfully",__func__);
++	}
++
++	dummy_proc_init(card);
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: What is the purpose of this api? and why dummy name?

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 15]
```diff
@@ -0,0 +1,1112 @@
...
++	return 0;
++
++register_fail:
++	printk("%s: Failure in	registering sound card",__func__);
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: This goto not needed. rather print this where actually occurred.

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 16]
```diff
@@ -0,0 +1,1112 @@
...
++	strlcpy(card->driver, "bt_audio_source", sizeof(card->driver));
++	strlcpy(card->shortname, card->driver, sizeof(card->shortname));
++	strlcpy(card->longname, "USB Bluetooth audio source",
++			sizeof(card->longname));
++
++	err = snd_card_register(card);
++	if (err){
++		goto register_fail;
++	} else {
++		printk("%s:Sound card registred sucecssfully",__func__);
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: No need to have else statement.

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 17]
```diff
@@ -0,0 +1,1112 @@
...
++
++	return err;
++}
++
++
++static void btusb_sco_disconnect(struct usb_interface *intf)
++{
++	struct btusb_data *data = usb_get_intfdata(intf);
++
++
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: remove extra lines.

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 18]
```diff
@@ -0,0 +1,1112 @@
...
++				(substream->runtime->frame_bits >> 3);
++	} else if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK) {
++		spin_lock_irqsave(&data->txlock, flags);
++
++		playback_hwptr_done = data->playback_hwptr_done;
++
++		if (data->playback_transfer_done >=
++				(MAX_PACKET_SIZE - SCO_HDR_SIZE)) {
++			data->playback_transfer_done -=
++					(MAX_PACKET_SIZE - SCO_HDR_SIZE);
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: Why this adjustment done here.. I believe it should be take care during processing itself.

File: android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch [hunk 19]
```diff
@@ -0,0 +1,1112 @@
...
++        memcpy(handle,(uint8_t*)buf,SCO_HANDLE_SIZE);
++        printk("value of handle = %x, %x",handle[0],handle[1]);
++        return 0;
++}
++
++static struct usb_driver btusb_sco_driver = {
++	.name		= "btusb_sco",
++	.probe		= btusb_sco_probe,
++	.unlocked_ioctl = btusb_sco_ioctl,
++	.disconnect	= btusb_sco_disconnect,
```
Comment (bpankajl) [inline android_p/google_diff/cel_apl/kernel/project-celadon/0006-Virtual-PCM-device-driver-for-enabling-BT-HFP.patch]: What if userspace does not call IOCTL to pass SCO handle? 

We need to handle this case. May be PCM open should fail if handle is not set properly or something else, but this case needs to be handled.

File: android_p/google_diff/cel_kbl/system/bt/0004-Disable-Wideband-Support.patch [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+   APPL_TRACE_DEBUG("%s", __func__);
+ 
+-  buf = "AT+BAC=1,2\r";
++  buf = "AT+BAC=1\r";
+ 
+   bta_hf_client_send_at(client_cb, BTA_HF_CLIENT_AT_BAC, buf, strlen(buf));
+ }
+-- 
+2.17.1
+
```
Comment (bpankajl) [inline android_p/google_diff/cel_kbl/system/bt/0004-Disable-Wideband-Support.patch]: I see lots of checkpatch errors on this driver patch...


>> total: 67 errors, 70 warnings, 1066 lines checked

Please fix applicable errors.

**Pull request merged**